pkg/pillar/devicenetwork/dhcpcd.go

```diff
@@ -211,16 +211,44 @@ func doDhcpClientInactivate(log *base.LogObject, nuc types.NetworkPortConfig) {
 		log.Functionf("doDhcpClientInactivate(%s) DT_NONE is a no-op\n",
 			nuc.IfName)
 	case types.DT_STATIC, types.DT_CLIENT:
-		extras := []string{}
-		if !dhcpcdCmd(log, "--release", extras, nuc.IfName, false) {
-			log.Errorf("doDhcpClientInactivate: release failed for %s\n",
-				nuc.IfName)
+		var extras []string
+		// Wait for a bit then give up
+		waitCount := 0
+		failed := false
+		for {
+			//it waits up to 10 seconds for exit https://github.com/NetworkConfiguration/dhcpcd/blob/dhcpcd-8.1.6/src/dhcpcd.c#L1950-L1957
+			if !dhcpcdCmd(log, "--release", extras, nuc.IfName, false) {
+				log.Errorf("doDhcpClientInactivate: release failed for %s", nuc.IfName)
+			}
+			if !dhcpcdExists(log, nuc.IfName) {
+				break
+			}
+			waitCount++
+			if waitCount >= 3 {
+				log.Errorf("doDhcpClientInactivate: dhcpcd %s still running, will kill it", nuc.IfName)
+				failed = true
+				break
+			}
+			log.Warnf("doDhcpClientInactivate: dhcpcd %s still running", nuc.IfName)
+			time.Sleep(1 * time.Second)
 		}
-		for dhcpcdExists(log, nuc.IfName) {
-			log.Warnf("dhcpcd %s still running", nuc.IfName)
+		if !failed {
+			log.Functionf("dhcpcd %s gone", nuc.IfName)
+			return
+		}
+		//kill dhcpcd process as a last resort
```

Comment:
How about trying dhcpcd -x <ifname> and see if that makes it go away?

Reply:
It may helps, but I cannot see in logs that in case of hung dhcpcd receive sigalrm, not sure that sigint will work.

Reply:
The release will make dhcpcd send a release message (and timeout and retry I assume) to tell the server it is no longer using the IP address. The exit would do that. 
But I don't know exactly why it is getting stuck - perhaps the release can take more than 30 second to complete? Is there an upper limits on its timeout?

Reply:
Sorry, it is hard to reverse-engineer dhcpcd for this case, but I cannot find any long waiting. I modified the command to be `--exit` for this case.

Reply:
I might make sense to add some instrumentation (timestamps + logs) so that we can see how long time the --release and --exit takes. It makes sense to try the --release for 10 seconds and then hopefully an exit will work (and if it does not then we need to understand why things are stuck).

Seems like the nim.touch watchdogs started happening recently based on searching in alpha Kibana. So something changed somewhere?

Reply:
I put elapsed time in logs.